File: lanscape-agent/internal/agent/session.go
package agent

import (
	"context"
	"log/slog"
)

// BrowserSession represents a single browser connection with its own WebRTC and signaling
type BrowserSession struct {
	webrtc    *WebRTCManager
	signaling *SignalingClient
	bridge    *Bridge
	logger    *slog.Logger
}

// NewBrowserSession creates a new browser session with its own WebRTC and signaling
func NewBrowserSession(signalingURL, topic string, tailscaleInfo *TailscaleInfo, logger *slog.Logger) (*BrowserSession, error) {
	// Create WebRTC manager for this session
	webrtc, err := NewWebRTCManager(tailscaleInfo, logger)
	if err != nil {
		return nil, err
	}

	// Create signaling client for this session (needed for bridge)
	signaling := NewSignalingClient(signalingURL, topic, webrtc, logger)

	// Create bridge
	bridge := NewBridge(webrtc, logger)
	
	// Set up signaling callback to send welcome to browser when received
	signaling.SetOnWelcome(func(selfID string) {
		bridge.sendWelcome(selfID)
	})

	// Set up ICE candidate callback
	webrtc.SetOnICECandidate(func(peerID string, candidate interface{}) {
		if candidate != nil {
			signaling.sendICECandidate(peerID, candidate)
		}
	})

	session := &BrowserSession{
		webrtc:    webrtc,
		signaling: signaling,
		bridge:    bridge,
		logger:    logger,
	}

	return session, nil
}

// Connect connects to the signaling server
func (s *BrowserSession) Connect() error {
	return s.signaling.Connect()
}

// Disconnect disconnects from signaling and closes all peer connections
func (s *BrowserSession) Disconnect() {
	s.signaling.Disconnect()
	s.webrtc.CloseAll()
}

// GetBridge returns the bridge for this session
func (s *BrowserSession) GetBridge() *Bridge {
	return s.bridge
}

// GetSelfID returns the self peer ID from signaling
func (s *BrowserSession) GetSelfID() string {
	return s.signaling.GetSelfID()
}

// Stop stops the session
func (s *BrowserSession) Stop(ctx context.Context) error {
	s.Disconnect()
	return nil
}

File: packages/tauri/src-tauri/agent/src/session.rs
use crate::bridge::Bridge;
use crate::signaling::SignalingClient;
use crate::tailscale::TailscaleInfo;
use crate::webrtc::{WebRTCManager, RTCIceCandidate};
use anyhow::Result;
use std::sync::Arc;
use tracing::info;

/// BrowserSession represents a single browser connection with its own WebRTC and signaling
pub struct BrowserSession {
    webrtc: Arc<WebRTCManager>,
    signaling: Arc<SignalingClient>,
    bridge: Arc<Bridge>,
}

impl BrowserSession {
    /// Create a new browser session with its own WebRTC and signaling
    pub async fn new(
        signaling_url: String,
        topic: String,
        tailscale_info: Option<TailscaleInfo>,
    ) -> Result<Self> {
        // Create WebRTC manager for this session
        let webrtc = Arc::new(WebRTCManager::new(tailscale_info)?);

        // Create signaling client for this session
        let signaling = Arc::new(SignalingClient::new(
            signaling_url.clone(),
            topic.clone(),
            webrtc.clone(),
        ));

        // Create bridge
        let bridge = Arc::new(Bridge::new(webrtc.clone()));

        // Set up signaling callback to send welcome to browser when received
        let bridge_welcome = bridge.clone();
        signaling
            .set_on_welcome(Box::new(move |self_id: String| {
                let bridge = bridge_welcome.clone();
                tokio::spawn(async move {
                    bridge.send_welcome(self_id).await;
                });
            }))
            .await;

        // Set up ICE candidate callback
        let signaling_ice = signaling.clone();
        webrtc
            .set_on_ice_candidate(Box::new(move |peer_id: String, candidate: RTCIceCandidate| {
                let signaling = signaling_ice.clone();
                Box::pin(async move {
                    if let Err(e) = signaling.send_ice_candidate(&peer_id, candidate).await {
                        tracing::warn!("Failed to send ICE candidate: peer={}, error={}", peer_id, e);
                    }
                })
            }))
            .await;

        Ok(Self {
            webrtc,
            signaling,
            bridge,
        })
    }

    /// Connect to the signaling server
    pub async fn connect(&self) -> Result<()> {
        self.signaling.connect().await
    }

    /// Disconnect from signaling and close all peer connections
    pub async fn disconnect(&self) {
        self.signaling.disconnect().await;
        self.webrtc.close_all().await;
    }

    /// Get the bridge for this session
    pub fn get_bridge(&self) -> Arc<Bridge> {
        self.bridge.clone()
    }

    /// Get self peer ID from signaling
    pub async fn get_self_id(&self) -> Option<String> {
        self.signaling.get_self_id().await
    }

    /// Stop the session
    pub async fn stop(&self) -> Result<()> {
        self.disconnect().await;
        Ok(())
    }
}

